tools/schema_loader.cc

```diff
@@ -318,6 +323,169 @@ std::vector<schema_ptr> do_load_schemas(std::string_view schema_str) {
+        auto table_name = de.name.substr(0, dash_pos);
+
+        auto it = boost::find_if(schemas, [&] (const schema_ptr& s) {
+            return s->cf_name() == table_name;
+        });
+
+        if (it != schemas.end()) {
+            if (!de.type) {
+                throw std::runtime_error(fmt::format("failed loading schema tables from {}: keyspace directory entry {} has unrecognized type", scylla_data_path.native(), de.name));
+            } else if (*de.type != directory_entry_type::directory) {
+                throw std::runtime_error(fmt::format("failed loading schema tables from {}: keyspace directory entry {} has unrecognized type {}", scylla_data_path.native(), de.name, static_cast<int>(*de.type)));
+            }
+            auto s = *it;
+            schema_table_table_dir[s] = de.name;
+        }
+        return make_ready_future<>();
+    }).done().get();
+
+    if (schema_table_table_dir.size() != schemas.size()) {
+        throw std::runtime_error(fmt::format("failed loading schema tables from {}: couldn't find table directory for all require schema tables", scylla_data_path.native()));
+    }
+
+    return schema_table_table_dir;
+}
+
+schema_ptr do_load_schema_from_schema_tables(std::filesystem::path scylla_data_path, std::string_view keyspace, std::string_view table) {
+    reader_concurrency_semaphore rcs_sem(reader_concurrency_semaphore::no_limits{}, __FUNCTION__);
+    auto stop_semaphore = deferred_stop(rcs_sem);
+    db::nop_large_data_handler large_data_handler;
+    db::config dbcfg;
+    gms::feature_service feature_service(gms::feature_config_from_db_config(dbcfg));
+    cache_tracker tracker;
+    dbcfg.host_id = locator::host_id::create_random_id();
+    sstables::directory_semaphore dir_sem(1);
+    sstables::sstables_manager sst_man(large_data_handler, dbcfg, feature_service, tracker, memory::stats().total_memory(), dir_sem);
+    auto close_sst_man = deferred_close(sst_man);
+
+    auto schema_table_table_dir = get_schema_table_directories(scylla_data_path);
+    auto schema_tables_path = scylla_data_path / db::schema_tables::NAME;
+
+    auto do_load = [&] (const schema_ptr s) {
+        return read_schema_table_mutation(
+                sst_man,
+                schema_tables_path / schema_table_table_dir[s],
+                s,
+                rcs_sem.make_tracking_only_permit(s.get(), "schema_mutation", db::no_timeout),
+                keyspace,
+                {table});
+    };
+    mutation tables = do_load(db::schema_tables::tables());
+    mutation columns = do_load(db::schema_tables::columns());
+    mutation view_virtual_columns = do_load(db::schema_tables::view_virtual_columns());
+    mutation computed_columns = do_load(db::schema_tables::computed_columns());
+    mutation indexes = do_load(db::schema_tables::indexes());
+    mutation dropped_columns = do_load(db::schema_tables::dropped_columns());
+    mutation scylla_tables = do_load(db::schema_tables::scylla_tables());
```

Comment:
Can't we reuse code from schema_tables.cc? This looks like a lot of duplication.

Reply:
We do reuse the schema_tables code actually. Note how below I call schema tables method to actually create the schema from the raw data. This is just preparing the arguments for that method.

Or you mean reuse the code which reads the schema from disk as a whole? I thought about that but I decided against it. The schema tables code does queries against the database to read the schema tables contents from disk. For this to work I'd have to bootstrap a minimal scylla environment, something like `cql_test_env`. This is entirely possible but there is one blocker for me: one of the main requirement I set for myself when writing `scylla sstable` is that it should be strictly read-only unless explicitly instructed to mutate/write an sstable. `cql_test_env` currently acts like it owns the place and will happily proceed compacting/resharding/reshaping sstables on disk as a normal scylla node would. I could try to hack in a read-only mode but I'm afraid I could miss some corner case, or the read-only mode could break and we wouldn't notice, and then a tool which is only meant to read sstables, would suddenly start compacting/resharding/reshaping or even deleting them. So I wrote my own code instead, which I know will not go rouge and attempt do mutate the sstable set.

Reply:
I don't think we should have query_processor running.

But, perhaps we can abstract its use to one noncopyable_function, and supply that to those functions.

Probably, the way to go is to make database::parse_system_tables() a free function that accepts this noncopyable_function. Though it's lot of work, since parse_system_tables() mutates the database object.

Reply:
It would involve, perhaps, working with data_dictionary to store the metadata rather directly using the database object. Then both the data_dictionary backed by the database class and your fake one would work.

Please look if it's feasible, but don't go overboard, a huge refactoring isn't worth it to remove a small duplication.

Reply:
This is how schema tables does it: https://github.com/scylladb/scylladb/blob/e3dc60286c36de26a1870377286c195de2fee5f8/db/schema_tables.cc#L2825-L2836

The method it calls is very similar to my `do_load` method, it prepares range/slice then calls a query on storage proxy, where I create a reader and read a mutation. I don't think this is worth trying to be made more generic, so my code can use it too.